Allow full length GPG key fingerprints.

WolverineFan opened the pull request:
This is a rebased version of rfkrocktk's patch #354 with added tests and several bug fixes found while running said tests.  I think this should be good to go.

daenney commented on spec/defines/key_spec.rb:
```diff
@@ -2,7 +2,8 @@
 describe 'apt::key', :type => :define do
   let(:facts) { { :lsbdistid => 'Debian' } }
-  GPG_KEY_ID = '4BD6EC30'
+  #GPG_KEY_ID = '4BD6EC30'
```
The comments serve no purpose, just remove them. We've got Git's history powers if we ever wanted to know what this line looked like in the past.

rfkrocktk replied:
Yes, this is a major pet peeve. Commented code adds _nothing_ unless its an example. Having Git means there's should be no such thing as "I'll comment this out in case I need it later."

WolverineFan replied:
I almost did, but it was late and I'd spent a couple hours rebasing/reformatting :)  I'll remove them.

daenney commented on spec/defines/source_spec.rb:
```diff
@@ -1,7 +1,9 @@
 
 describe 'apt::source', :type => :define do
-  GPG_KEY_ID = '4BD6EC30'
+  #GPG_KEY_ID = '4BD6EC30'
```
idem

daenney commented on spec/acceptance/apt_key_provider_spec.rb:
```diff
@@ -42,7 +49,7 @@
           apply_manifest(pp, :catch_failures => true)
           apply_manifest(pp, :catch_changes => true)
-          shell("apt-key list | grep #{PUPPETLABS_GPG_KEY_ID}")
+          shell("apt-key adv --list-keys --with-colons --fingerprint | grep #{PUPPETLABS_GPG_KEY_FINGERPRINT}")
```
In the provider you created a `cli_args` so we didn't have to repeat ourselves all over the place. Could you repeat that for the tests too?

cyberious replied:
This really feels more like a shared_examples scenario as well.

WolverineFan replied:
What is a shared_examples scenario?

WolverineFan replied:
I found the docs for shared_examples: http://www.relishapp.com/rspec/rspec-core/v/2-11/docs/example-groups/shared-examples

I'm going to leave such a refactoring to someone who knows the tools better.  I'll change the shell commands to use a variable though.

daenney commented on spec/acceptance/apt_key_provider_spec.rb:
```diff
@@ -67,25 +74,25 @@
-              --recv-keys #{CENTOS_GPG_KEY_ID}")
-        shell("apt-key list | grep #{CENTOS_GPG_KEY_ID}")
+              --recv-keys #{CENTOS_GPG_KEY_FINGERPRINT}")
+        shell("apt-key adv --list-keys --with-colons --fingerprint | grep #{CENTOS_GPG_KEY_FINGERPRINT}")
```
idem

daenney commented on spec/acceptance/apt_key_provider_spec.rb:
```diff
@@ -67,25 +74,25 @@
         apply_manifest(pp, :catch_failures => true)
 
-        shell("apt-key list | grep #{CENTOS_GPG_KEY_ID}",
+        shell("apt-key adv --list-keys --with-colons --fingerprint | grep #{CENTOS_GPG_KEY_FINGERPRINT}",
```
idem

daenney commented on spec/acceptance/apt_key_provider_spec.rb:
```diff
@@ -93,21 +100,21 @@
-              --recv-keys #{PUPPETLABS_GPG_LONG_KEY_ID}")
-        shell("apt-key list | grep #{PUPPETLABS_GPG_KEY_ID}")
+              --recv-keys #{PUPPETLABS_GPG_KEY_LONG_ID}")
+        shell("apt-key adv --list-keys --with-colons --fingerprint | grep #{PUPPETLABS_GPG_KEY_FINGERPRINT}")
```
idem

daenney commented on spec/acceptance/apt_key_provider_spec.rb:
```diff
@@ -93,21 +100,21 @@
         apply_manifest(pp, :catch_failures => true)
 
-        shell("apt-key list | grep #{PUPPETLABS_GPG_KEY_ID}",
+        shell("apt-key adv --list-keys --with-colons --fingerprint | grep #{PUPPETLABS_GPG_KEY_FINGERPRINT}",
```
idem

daenney commented on spec/acceptance/apt_key_provider_spec.rb:
```diff
@@ -185,7 +192,7 @@
         apply_manifest(pp, :catch_failures => true)
         apply_manifest(pp, :catch_failures => true)
-        shell("apt-key list | grep #{PUPPETLABS_GPG_KEY_ID}")
+        shell("apt-key adv --list-keys --with-colons --fingerprint | grep #{PUPPETLABS_GPG_KEY_FINGERPRINT}")
```
idem

daenney commented on spec/acceptance/apt_key_provider_spec.rb:
```diff
@@ -219,7 +226,7 @@
         apply_manifest(pp, :catch_failures => true)
         apply_manifest(pp, :catch_failures => true)
-        shell("apt-key list | grep #{PUPPETLABS_GPG_KEY_ID}")
+        shell("apt-key adv --list-keys --with-colons --fingerprint | grep #{PUPPETLABS_GPG_KEY_FINGERPRINT}")
```
idem

daenney commented on spec/acceptance/apt_key_provider_spec.rb:
```diff
@@ -235,7 +242,7 @@
         apply_manifest(pp, :catch_failures => true)
         apply_manifest(pp, :catch_failures => true)
-        shell("apt-key list | grep #{PUPPETLABS_GPG_KEY_ID}")
+        shell("apt-key adv --list-keys --with-colons --fingerprint | grep #{PUPPETLABS_GPG_KEY_FINGERPRINT}")
```
idem

daenney commented on spec/acceptance/apt_key_provider_spec.rb:
```diff
@@ -285,13 +292,13 @@
         apply_manifest(pp, :catch_failures => true)
         apply_manifest(pp, :catch_failures => true)
-        shell("apt-key list | grep #{PUPPETLABS_GPG_KEY_ID}")
+        shell("apt-key adv --list-keys --with-colons --fingerprint | grep #{PUPPETLABS_GPG_KEY_FINGERPRINT}")
```
idem

daenney commented on spec/acceptance/apt_key_provider_spec.rb:
```diff
@@ -334,13 +341,13 @@
         apply_manifest(pp, :catch_failures => true)
         apply_manifest(pp, :catch_failures => true)
-        shell("apt-key list | grep #{CENTOS_GPG_KEY_ID}")
+        shell("apt-key adv --list-keys --with-colons --fingerprint | grep #{CENTOS_GPG_KEY_FINGERPRINT}")
```
idem

daenney commented on spec/acceptance/apt_key_provider_spec.rb:
```diff
@@ -378,7 +385,7 @@
         apply_manifest(pp, :catch_failures => true)
         apply_manifest(pp, :catch_failures => true)
-        shell("apt-key list | grep #{PUPPETLABS_GPG_KEY_ID}")
+        shell("apt-key adv --list-keys --with-colons --fingerprint | grep #{PUPPETLABS_GPG_KEY_FINGERPRINT}")
```
idem

daenney commented on spec/acceptance/apt_key_provider_spec.rb:
```diff
@@ -431,7 +438,7 @@
         apply_manifest(pp, :catch_failures => true)
         apply_manifest(pp, :catch_failures => true)
-        shell("apt-key list | grep #{PUPPETLABS_GPG_KEY_ID}")
+        shell("apt-key adv --list-keys --with-colons --fingerprint | grep #{PUPPETLABS_GPG_KEY_FINGERPRINT}")
```
idem

daenney commented on spec/acceptance/apt_key_provider_spec.rb:
```diff
@@ -488,19 +495,19 @@
         apply_manifest(pp, :catch_failures => true)
         apply_manifest(pp, :catch_failures => true)
-        shell("apt-key list | grep #{PUPPETLABS_GPG_KEY_ID}")
+        shell("apt-key adv --list-keys --with-colons --fingerprint | grep #{PUPPETLABS_GPG_KEY_FINGERPRINT}")
```
idem

daenney commented on lib/puppet/provider/apt_key/apt_key.rb:
```diff
@@ -57,59 +68,46 @@ def self.instances
+      return_hash[:key_type] = :rsa
+    when "17"
+      return_hash[:key_type] = :dsa
+    end
```
Could you add:
* 18: ECC
* 19: ECDSA

rfkrocktk replied:
It would be nice to have these, but IIRC, they're really rare. Elliptic Curve algorithms _should_ be more secure with smaller key sizes, but you _really_ have to go out of your way to use them when creating the keys.

daenney replied:
Yes but they're up and coming, I've already seen these keys in the wild. It's not that much work to add two case statements.

cyberious replied:
Key ID was always the short key, so I am fine with us renaming, as long as we create the resource with the id and on prefetch we determine which length to match.

daenney commented on lib/puppet/provider/apt_key/apt_key.rb:
```diff
@@ -57,59 +68,46 @@ def self.instances
-      :key_type    => line_array[2],
-      :key_created => line_array[4],
-      :key_expiry  => nil,
+      :key_fingerprint => fingerprint,
```
Shouldn't we keep the `:key_id`?

rfkrocktk replied:
Also, shouldn't we keep the type, size, created, and expiry as well? More information would be better than less.

daenney replied:
Yeah, good point. Especially since those fields are still defined on the type. We're losing information here.

cyberious replied:
I think we can remove key_id as it is only the hash and holds no meaning seeing as we are matching on short / long or fingerprint.  ID is really a non specific value.

WolverineFan replied:
@cyberious is correct.  key_id was was renamed key_short.  The other values are below here, nothing was removed.

WolverineFan commented:
Updated based on feedback and rebased on current master again.  Can someone take a look at my #403 pull request too?  The apt_has_updates fact doesn't work at all at the moment.

daenney commented:
@WolverineFan Don't do that please, asking to push through another PR through this one unless they're related. We'll get to it, but we can't 24/7 our lives on this.

daenney commented:
I've merged this so we don't sit on it any longer and the PR diverges again. It seems good, tests pass, lets see what happens next.

rfkrocktk commented:
Great job everyone!

cyberious commented on lib/puppet/provider/apt_key/apt_key.rb:
```diff
@@ -41,14 +49,17 @@ def self.instances
-        :type    => line_hash[:key_type] == 'R' ? :rsa : :dsa,
-        :created => line_hash[:key_created]
+        :name        => line_hash[:key_fingerprint],
+        :id          => line_hash[:key_fingerprint],
```
Is this really what we wanted, we are not reporting back the id as a fingerprint, previously this was the short.

WolverineFan replied:
True, but short is bad (not nearly unique enough).  So I figured it had to change to long or fingerprint and chose fingerprint.  It's a simple change back to short if it causes some breaking behavior (though I couldn't find anything that uses it).

rfkrocktk replied:
This might be a candidate for a major version bump.
On Jan 14, 2015 7:31 PM, "WolverineFan" <notifications@github.com> wrote:

> In lib/puppet/provider/apt_key/apt_key.rb
> <https://github.com/puppetlabs/puppetlabs-apt/pull/404#discussion-diff-22988859>
> :
>
> > @@ -41,14 +49,17 @@ def self.instances
> >        end
> >
> >        new(
> > -        :name    => line_hash[:key_id],
> > -        :id      => line_hash[:key_id],
> > -        :ensure  => :present,
> > -        :expired => expired,
> > -        :expiry  => line_hash[:key_expiry],
> > -        :size    => line_hash[:key_size],
> > -        :type    => line_hash[:key_type] == 'R' ? :rsa : :dsa,
> > -        :created => line_hash[:key_created]
> > +        :name        => line_hash[:key_fingerprint],
> > +        :id          => line_hash[:key_fingerprint],
>
> True, but short is bad (not nearly unique enough). So I figured it had to
> change to long or fingerprint and chose fingerprint. It's a simple change
> back to short if it causes some breaking behavior (though I couldn't find
> anything that uses it).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetlabs-apt/pull/404/files#r22988859>.
>

rfkrocktk replied:
Also, the build is broken.
On Jan 14, 2015 7:40 PM, "Naftuli Tzvi Kay" <rfkrocktk@gmail.com> wrote:

> This might be a candidate for a major version bump.
> On Jan 14, 2015 7:31 PM, "WolverineFan" <notifications@github.com> wrote:
>
>> In lib/puppet/provider/apt_key/apt_key.rb
>> <https://github.com/puppetlabs/puppetlabs-apt/pull/404#discussion-diff-22988859>
>> :
>>
>> > @@ -41,14 +49,17 @@ def self.instances
>> >        end
>> >
>> >        new(
>> > -        :name    => line_hash[:key_id],
>> > -        :id      => line_hash[:key_id],
>> > -        :ensure  => :present,
>> > -        :expired => expired,
>> > -        :expiry  => line_hash[:key_expiry],
>> > -        :size    => line_hash[:key_size],
>> > -        :type    => line_hash[:key_type] == 'R' ? :rsa : :dsa,
>> > -        :created => line_hash[:key_created]
>> > +        :name        => line_hash[:key_fingerprint],
>> > +        :id          => line_hash[:key_fingerprint],
>>
>> True, but short is bad (not nearly unique enough). So I figured it had to
>> change to long or fingerprint and chose fingerprint. It's a simple change
>> back to short if it causes some breaking behavior (though I couldn't find
>> anything that uses it).
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/puppetlabs/puppetlabs-apt/pull/404/files#r22988859>.
>>
>

WolverineFan commented:
I'm pretty the failures aren't because of this patch.  My guess is the failure is in the test framework because it is consistently reporting getting nil for things (ironically the failures are in the code I fixed in my other patch, but my version doesn't work either).  Also, running the all tests on my box against current master works fine.

cyberious commented:
No the failures were directly part of how we were deleting the keys.  Previously the failures were only b/c there were duplicate keys and we would not delete the expired key but just the first one apt-key found.  Now it wasn't deleting them at all.

WolverineFan commented:
That's odd.  The tests have several checks for key deletion, so I'm very surprised if that broke as part of this patch.  Plus the tests that are listed as failing are all apt_has_updates checks that refer to unexpected results.

WolverineFan commented:
Just realized I can pretty much guarantee this patch didn't break the tests:  Nothing was committed to master between my last rebase and when it was merged.  My last rebase passed all tests.  So the exact same commit SHA passed once and failed the second time.  That points to a testing environment issue.

(By the way, I'm not implying there aren't problems with the patch, just that the test failures aren't because of it)

rfkrocktk commented:
The builds were passing until it was merged. I'm not saying it was the
patch, but it's quite an interesting coincidence.


On Wed, Jan 14, 2015 at 9:41 PM, WolverineFan <notifications@github.com>
wrote:

> Just realized I can pretty much guarantee this patch didn't break the
> tests: Nothing was committed to master between my last rebase and when it
> was merged. My last rebase passed all tests. So the exact same commit SHA
> passed once and failed the second time. That points to a testing
> environment issue.
>
> (By the way, I'm not implying there aren't problems with the patch, just
> that the test failures aren't because of it)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetlabs-apt/pull/404#issuecomment-70042369>
> .
>

cyberious commented:
It isn't just the travis tests that I am referencing but also the acceptance tests we run internally.  But nevertheless they are working now.  Also given the community feedback I have created a PR #410 which changes the delete to use fingerprints as well.

rfkrocktk commented:
Now that this is involving changes to other places in the codebase, I'm even more in favor of a major version bump.

cyberious commented:
@rfkrocktk the changes that have occurred do not warrant a major version bump.  I will look at what has occurred outside of this scope of these PR's

rfkrocktk commented:
Aren't they backwards-breaking changes?
On Jan 15, 2015 9:08 AM, "Travis Fields" <notifications@github.com> wrote:

> @rfkrocktk <https://github.com/rfkrocktk> the changes that have occurred
> do not warrant a major version bump. I will look at what has occurred
> outside of this scope of these PR's
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetlabs-apt/pull/404#issuecomment-70120779>
> .
>

cyberious commented:
How so?  The API now takes 8, 16 or 40, we have added.  I don't see it as a breaking change, we have exactly the same behavior just different mechanism of doing it.  My only concern was that the puppet resource apt_key now does not report the namevar as short key.

rfkrocktk commented:
My apologies, I thought that there were breaking changes, nevermind.
On Jan 15, 2015 10:10 AM, "Travis Fields" <notifications@github.com> wrote:

> How so? The API now takes 8, 16 or 40, we have added. I don't see it as a
> breaking change, we have exactly the same behavior just different mechanism
> of doing it. My only concern was that the puppet resource apt_key now does
> not report the namevar as short key.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetlabs-apt/pull/404#issuecomment-70132380>
> .
>
